### PR TITLE
Serialise function types and argument operands.

### PR DESCRIPTION
**Pull request by vext01**
(yk PR ~~coming soon~~: https://github.com/ykjit/yk/pull/881)

### CODE REVIEW

File: llvm/lib/YkIR/YkIRWriter.cpp [hunk 1]
```diff
@@ -133,8 +135,17 @@ class YkIRWriter {
     if (Found != Types.end()) {
       return std::distance(Types.begin(), Found);
     }
+
+    // Not found. Assign it a type index.
     size_t Idx = Types.size();
     Types.push_back(Ty);
+
+    // If the newly-registered type has any sub-types, then assign them type
```
[the remark this answers is not in the document]
Comment (vext01) [reply llvm/lib/YkIR/YkIRWriter.cpp:143]: For example, a struct type is made up of other types. Same for arrays. It's re-used [LLVM terminology](https://llvm.org/doxygen/classllvm_1_1Type.html#a5ea992fd3e9da95e4fff1dca95f873c3)

The problem I'm solving here is, when we serialise types (which happens last), we cannot introduce any new types, or it'd amount to iterator invalidation.

I plan to do a follow up PR to either make it impossible to add new types during type serialisation, or at least have the serialiser crash if it detects it.

[the remark this answers is not in the document]
Comment (vext01) [reply llvm/lib/YkIR/YkIRWriter.cpp:143]: Ah we raced. Yes, I can re-word it.

[the remark this answers is not in the document]
Comment (vext01) [reply llvm/lib/YkIR/YkIRWriter.cpp]: d4c6ebf1fe96 avoids the loaded terminology.

**Comment by ltratt** [comment]
Please squash.

**Comment by vext01** [comment]
Ugh. My apologies. Fixed. Squash?

**Comment by ltratt** [comment]
Please squash.

**Pull request merged**